File: game/scripts/vscripts/heroes/meepo/divided_we_stand.ts
// For this to work you need to override the following lines in your lua file
// GameRules.GetGameModeEntity(GameRules).SetExecuteOrderFilter(GameRules.GetGameModeEntity(GameRules),function(f,filterTable)
// GameRules.GetGameModeEntity(GameRules).SetModifyExperienceFilter(GameRules.GetGameModeEntity(GameRules),function(f,filterTable)

// Known issue: Power treads not moving along, there is no way to get their state.
// Multiple boots may be included in the networth
LinkLuaModifier("modifier_meepo_divided_we_stand_lua","heroes/meepo/divided_we_stand.lua",LuaModifierType.LUA_MODIFIER_MOTION_NONE)
LinkLuaModifier("modifier_meepo_divided_we_stand_death_lua","heroes/meepo/divided_we_stand.lua",LuaModifierType.LUA_MODIFIER_MOTION_NONE)

class meepo_divided_we_stand_lua extends CDOTA_Ability_Lua {
	isScepterUpgraded: boolean = false

	
	OnUpgrade() {
		let caster = <CDOTA_BaseNPC_Hero>this.GetCaster()
		let PID = caster.GetPlayerOwnerID()
		let mainMeepo = PlayerResource.GetSelectedHeroEntity(PID)
		let list:CDOTA_BaseNPC_Hero[] = mainMeepo.meepoList  || []
		if (caster != mainMeepo) {
			return null
		}
		if (!mainMeepo.meepoList) {
			list.push(mainMeepo)
			mainMeepo.meepoList = list
			mainMeepo.AddNewModifier(mainMeepo,this,"modifier_meepo_divided_we_stand_lua",{})
		}

		let newMeepo = <CDOTA_BaseNPC_Hero>CreateUnitByName(
			caster.GetUnitName(),
			mainMeepo.GetAbsOrigin(),
			true,mainMeepo,
			mainMeepo.GetPlayerOwner(),
			mainMeepo.GetTeamNumber()
		)
		newMeepo.SetControllableByPlayer(PID,false)
		newMeepo.SetOwner(caster.GetOwner())
		newMeepo.AddNewModifier(mainMeepo,this,"modifier_phased",{duration:0.1})
		let ability = newMeepo.FindAbilityByName(this.GetAbilityName())
		newMeepo.AddNewModifier(mainMeepo,this,"modifier_meepo_divided_we_stand_lua",{})
		list = mainMeepo.meepoList
		list.push(newMeepo)
		mainMeepo.meepoList = list
	}

	OnInventoryContentsChanged() {
		// The scepter part
		if (!this.isScepterUpgraded && this.GetCaster().HasScepter()) {
			for (let i=0 ; i<=5 ; i++) {
				let item:CDOTA_Item = this.GetCaster().GetItemInSlot(i)
				if (item && item.GetAbilityName() == "item_ultimate_scepter") {
					item.SetDroppable(false)
					item.SetSellable(false)
					item.SetCanBeUsedOutOfInventory(false)
					this.isScepterUpgraded = true
				}
			}
			this.OnUpgrade()
		}
	}
}

// Modifier stuff

class modifier_meepo_divided_we_stand_lua extends CDOTA_Modifier_Lua {
	IsHidden() {return true}
	IsPermanent() {return true}
	boots_list:string[]

	DeclareFunctions() {
		return [
			modifierfunction.MODIFIER_EVENT_ON_ORDER,
			modifierfunction.MODIFIER_EVENT_ON_DEATH,
			modifierfunction.MODIFIER_EVENT_ON_RESPAWN,
			modifierfunction.MODIFIER_EVENT_ON_TAKEDAMAGE,
		]
	}


	// Stats sync with client on refresh and no other syncing needed.(?)
	OnCreated() {
		if (IsServer()) {
			this.StartIntervalThink(FrameTime())
		}
	}

	OnIntervalThink() {
		let meepo = <CDOTA_BaseNPC_Hero>this.GetParent()
		let mainMeepo = <CDOTA_BaseNPC_Hero>this.GetCaster()
		
		let ability: CDOTABaseAbility = this.GetAbility()
		// Loop through abilities to sync levels
		if (mainMeepo != meepo) {
			let boots:string[] = [
				"item_travel_boots2",
				"item_travel_boots",
				"item_guardian_greaves",
				"item_power_treads",
				"item_arcane_boots",
				"item_phase_boots",
				"item_tranquil_boots",
				"item_boots"
			]
			// Get the highest ranked boots
			let item = ""
			for (let name of boots) {
				if (item == "") {
					for (let j=0;j<=5;j++) {
						let it = mainMeepo.GetItemInSlot(j)
						if (it && name == it.GetAbilityName()) {
							item = name
						}
					}
				} else {
					break
				}
			}
			// Remove previous item if it exists and isn't the same
			// Add the best boots
			if (item != "") {
				if (meepo["item"]) { 
					if (meepo["item"] != item) {
						UTIL_Remove(meepo["itemHandle"])
						let itemHandle = meepo.AddItemByName(item)
						itemHandle.SetDroppable(false)
						itemHandle.SetSellable(false)
						itemHandle.SetCanBeUsedOutOfInventory(false)
						meepo["itemHandle"] = itemHandle
						meepo["item"] = item
					}
				} else {
					meepo["itemHandle"] = meepo.AddItemByName(item)
					meepo["item"] = item
				}
				
				// Remove all items from clones
			}  
			for (let j=0;j<=5;j++) {
				let itemToCheck:CDOTA_Item = meepo.GetItemInSlot(j)
				if (itemToCheck) {
					let name = itemToCheck.GetAbilityName()
					if (name != item) {
						UTIL_Remove(itemToCheck)
					}
				}
			}
			
			// Set base stats equal to main meepo stats
			meepo.SetBaseStrength(mainMeepo.GetStrength())
			meepo.SetBaseAgility(mainMeepo.GetAgility())
			meepo.SetBaseIntellect(mainMeepo.GetIntellect())
			meepo.CalculateStatBonus()
			
			while(meepo.GetLevel() < mainMeepo.GetLevel()) {
				meepo.AddExperience(10,1,false,false)
			}
		} else {
			LevelAbilitiesForAllMeepos(meepo)
		}
		
	}

	// Use this to only kill the main meepo and hide the others
	OnTakeDamage(keys:{inflictor:CDOTABaseAbility, unit:CDOTA_BaseNPC,target:CDOTA_BaseNPC}) {
		let mainMeepo = <CDOTA_BaseNPC_Hero>this.GetCaster()
		let parent = <CDOTA_BaseNPC_Hero>this.GetParent()
		if (keys.unit == parent) {
			if (parent.GetHealth() < 0) {
				if (parent != this.GetCaster()) {
					// Move the main hero to this location to grant exp etc, then move back in case of aegis/bloodstone
					let oldLocation = mainMeepo.GetAbsOrigin()
					mainMeepo.SetAbsOrigin(parent.GetAbsOrigin())
					mainMeepo.Kill(keys.inflictor,keys.attacker)
					mainMeepo.SetAbsOrigin(oldLocation)

					// Make sure the hero survives this
					parent.SetHealth(parent.GetMaxHealth())
				}
				for (let meepo of GetAllMeepos(mainMeepo)) {
					if (meepo != mainMeepo) {
						meepo.AddNewModifier(mainMeepo,this.GetAbility(),"modifier_meepo_divided_we_stand_death_lua",{})
					}
				}
			}
		}
	}

	OnRespawn(keys:{unit:CDOTA_BaseNPC_Hero}) {
		let parent = <CDOTA_BaseNPC_Hero> this.GetParent()
		let mainMeepo = <CDOTA_BaseNPC_Hero> this.GetCaster()
		if (keys.unit == parent && parent == PlayerResource.GetSelectedHeroEntity(this.GetParent().GetPlayerOwnerID())) {
			for (let meepo of GetAllMeepos(mainMeepo)) {
				if (meepo != mainMeepo) {  
					meepo.RemoveModifierByName("modifier_meepo_divided_we_stand_death_lua")
					meepo.RemoveNoDraw()
					FindClearSpaceForUnit(meepo,this.GetParent().GetAbsOrigin(),true)
					meepo.AddNewModifier(meepo, this.GetAbility(), "modifier_phased", {duration:0.1})
				}
			}
		}
	}
}
// Use this to fake deaths
class modifier_meepo_divided_we_stand_death_lua extends CDOTA_Modifier_Lua {
	IsPermanent() { return false}
	IsHidden() { return true }

	OnCreated() {
		if (IsServer()) {
			this.GetParent().StartGesture(GameActivity_t.ACT_DOTA_DIE)
			this.StartIntervalThink(1.5)
		}
	}

	OnIntervalThink() {
		let parent = this.GetParent()
		parent.RemoveGesture(GameActivity_t.ACT_DOTA_DIE)
		parent.AddNoDraw()
		if (parent.GetTeamNumber() == DOTATeam_t.DOTA_TEAM_GOODGUYS) {
			parent.SetAbsOrigin(Vector(-10000,-10000,0))
		} else {
			parent.SetAbsOrigin(Vector(10000,10000,0))
		}
		this.StartIntervalThink(-1) 
	}

	CheckState() {
		return {
			[modifierstate.MODIFIER_STATE_STUNNED]: true,
			[modifierstate.MODIFIER_STATE_UNSELECTABLE]:true,
			[modifierstate.MODIFIER_STATE_INVULNERABLE]:true,
			[modifierstate.MODIFIER_STATE_UNTARGETABLE]:true,
			[modifierstate.MODIFIER_STATE_OUT_OF_GAME]:true,
			[modifierstate.MODIFIER_STATE_NO_HEALTH_BAR]:true,
			[modifierstate.MODIFIER_STATE_NOT_ON_MINIMAP]:true,
		}
	}
}

// LevelAbilityForAllMeepos
// This can't be done in the onupgrade blocks because of talents
function LevelAbilitiesForAllMeepos(caster:CDOTA_BaseNPC_Hero): void {
	let PID = caster.GetPlayerOwnerID()
	let mainMeepo = PlayerResource.GetSelectedHeroEntity(PID)
	
	if (caster == mainMeepo) {
		for (let a=0;a <= caster.GetAbilityCount()-1;a++) {
			let ability: CDOTABaseAbility = caster.GetAbilityByIndex(a) 
			if (ability) {
				for (let meepo of GetAllMeepos(mainMeepo)) {
					let cloneAbility = meepo.FindAbilityByName(ability.GetAbilityName())
					if (ability.GetLevel() > cloneAbility.GetLevel()) {
						cloneAbility.SetLevel(ability.GetLevel())
						meepo.SetAbilityPoints(meepo.GetAbilityPoints()-1)
					} 
					if (ability.GetLevel() < cloneAbility.GetLevel()) {
						ability.SetLevel(cloneAbility.GetLevel())
						mainMeepo.SetAbilityPoints(mainMeepo.GetAbilityPoints()-1)
					}
				}
			}
		}
	}  
}

function GetAllMeepos(caster:CDOTA_BaseNPC_Hero):CDOTA_BaseNPC_Hero[] {
	if (caster.meepoList) { 
		return caster.meepoList
	} else {
		return [caster]
	}
}

function MeepoExperience(filterTable:table): table {
	let PID = <PlayerID> filterTable.player_id_const
	let reason = <EDOTA_ModifyXP_Reason> filterTable.reason_const 
	let experience = <number> filterTable.experience
	let hero = PlayerResource.GetSelectedHeroEntity(PID)
	if (hero && hero.HasAbility("meepo_divided_we_stand_lua") && reason != EDOTA_ModifyXP_Reason.DOTA_ModifyXP_Unspecified) {
		filterTable.experience = 0
		hero.AddExperience(experience,EDOTA_ModifyXP_Reason.DOTA_ModifyXP_Unspecified,true,false)
	}
	return filterTable
}

function MeepoOrderFilter(filterTable:table):boolean {
	let entindex_ability = <number>filterTable.entindex_ability
	let sequence_number_const = <number>filterTable.sequence_number_const   
	let queue = <0|1>filterTable.sequence_number_const
	let units = <number[]>filterTable.units
	let entindex_target = <number>filterTable.entindex_target   
	let position = Vector(filterTable.position_y,filterTable.position_y,filterTable.position_z)
	let order_type = <DotaUnitOrder_t>filterTable.order_type   
	let issuer_player_id_const = <PlayerID>filterTable.issuer_player_id_const
	let ability = EntIndexToHScript(entindex_ability)
	let target = EntIndexToHScript(entindex_target)


	for (let entindex_unit of units) {
		let unit = EntIndexToHScript(entindex_unit)
		// Cancel picking up items as clone
		if (unit.HasModifier("modifier_meepo_divided_we_stand_lua") && unit != PlayerResource.GetSelectedHeroEntity(unit.GetPlayerOwnerID())) {
			if (order_type == DotaUnitOrder_t.DOTA_UNIT_ORDER_PICKUP_ITEM) {
				return false
			}
		}
		// Prevent giving items to meepo
		if (target && target.HasModifier && target.HasModifier("modifier_meepo_divided_we_stand_lua") && target != PlayerResource.GetSelectedHeroEntity(target.GetPlayerOwnerID())) {
			if (order_type == DotaUnitOrder_t.DOTA_UNIT_ORDER_GIVE_ITEM) {
				return false
			}
		}
	}
	return true
}


if (IsServer()) {
	// If you are using an order filter yourself already, copy the meepo related line in there
	GameRules.GetGameModeEntity().SetExecuteOrderFilter(
		function(filterTable):boolean {
			//let meepo = require('heroes/meepo/divided_we_stand')
			if (MeepoOrderFilter) {
				if (!MeepoOrderFilter(filterTable)) {
					return false
				}
			}
			return true
		},
		GameRules.GetGameModeEntity()
	)

	// If you are using an experience filter yourself already, copy the meepo related line in there
	GameRules.GetGameModeEntity().SetModifyExperienceFilter(
		function(filterTable):boolean {
			//let meepo = require('heroes/meepo/divided_we_stand')
			if (MeepoExperience) {
				filterTable = MeepoExperience(filterTable)
			}
			return true
		},
		GameRules.GetGameModeEntity()
	)
}

File: game/scripts/vscripts/heroes/meepo/divided_we_stand.lua
LinkLuaModifier("modifier_meepo_divided_we_stand_lua","heroes/meepo/divided_we_stand.lua",LUA_MODIFIER_MOTION_NONE)
LinkLuaModifier("modifier_meepo_divided_we_stand_death_lua","heroes/meepo/divided_we_stand.lua",LUA_MODIFIER_MOTION_NONE)
meepo_divided_we_stand_lua = class({})
function meepo_divided_we_stand_lua.constructor(self)
    self.isScepterUpgraded = false
end
function meepo_divided_we_stand_lua.OnUpgrade(self)
    local caster = self.GetCaster(self)
    local PID = caster.GetPlayerOwnerID(caster)
    local mainMeepo = PlayerResource.GetSelectedHeroEntity(PlayerResource,PID)
    local list = mainMeepo.meepoList or {}
    if caster~=mainMeepo then
        return nil
    end
    if not mainMeepo.meepoList then
        table.insert(list, mainMeepo)
        mainMeepo.meepoList=list
        mainMeepo.AddNewModifier(mainMeepo,mainMeepo,self,"modifier_meepo_divided_we_stand_lua",{})
    end
    local newMeepo = CreateUnitByName(caster.GetUnitName(caster),mainMeepo.GetAbsOrigin(mainMeepo),true,mainMeepo,mainMeepo.GetPlayerOwner(mainMeepo),mainMeepo.GetTeamNumber(mainMeepo))
    newMeepo.SetControllableByPlayer(newMeepo,PID,false)
    newMeepo.SetOwner(newMeepo,caster.GetOwner(caster))
    newMeepo.AddNewModifier(newMeepo,mainMeepo,self,"modifier_phased",{["duration"]=0.1})
    local ability = newMeepo.FindAbilityByName(newMeepo,self.GetAbilityName(self))
    newMeepo.AddNewModifier(newMeepo,mainMeepo,self,"modifier_meepo_divided_we_stand_lua",{})
    list=mainMeepo.meepoList
    table.insert(list, newMeepo)
    mainMeepo.meepoList=list
end
function meepo_divided_we_stand_lua.OnInventoryContentsChanged(self)
    if not self.isScepterUpgraded and self.GetCaster(self).HasScepter(self.GetCaster(self)) then
        for i=0,5,1 do
            local item = self.GetCaster(self).GetItemInSlot(self.GetCaster(self),i)
            if item and (item.GetAbilityName(item)=="item_ultimate_scepter") then
                item.SetDroppable(item,false)
                item.SetSellable(item,false)
                item.SetCanBeUsedOutOfInventory(item,false)
                self.isScepterUpgraded=true
            end
        end
        self.OnUpgrade(self)
    end
end
modifier_meepo_divided_we_stand_lua = class({})
function modifier_meepo_divided_we_stand_lua.IsHidden(self)
    return true
end
function modifier_meepo_divided_we_stand_lua.IsPermanent(self)
    return true
end
function modifier_meepo_divided_we_stand_lua.DeclareFunctions(self)
    return {MODIFIER_EVENT_ON_ORDER,MODIFIER_EVENT_ON_DEATH,MODIFIER_EVENT_ON_RESPAWN,MODIFIER_EVENT_ON_TAKEDAMAGE}
end
function modifier_meepo_divided_we_stand_lua.OnCreated(self)
    if IsServer() then
        self.StartIntervalThink(self,FrameTime())
    end
end
function modifier_meepo_divided_we_stand_lua.OnIntervalThink(self)
    local meepo = self.GetParent(self)
    local mainMeepo = self.GetCaster(self)
    local ability = self.GetAbility(self)
    if mainMeepo~=meepo then
        local boots = {"item_travel_boots2","item_travel_boots","item_guardian_greaves","item_power_treads","item_arcane_boots","item_phase_boots","item_tranquil_boots","item_boots"}
        local item = ""
        for _, name in pairs(boots) do
            if item=="" then
                for j=0,5,1 do
                    local it = mainMeepo.GetItemInSlot(mainMeepo,j)
                    if it and (name==it.GetAbilityName(it)) then
                        item=name
                    end
                end
            else
                break
            end
        end
        if item~="" then
            if meepo["item"] then
                if meepo["item"]~=item then
                    UTIL_Remove(meepo["itemHandle"])
                    local itemHandle = meepo.AddItemByName(meepo,item)
                    itemHandle.SetDroppable(itemHandle,false)
                    itemHandle.SetSellable(itemHandle,false)
                    itemHandle.SetCanBeUsedOutOfInventory(itemHandle,false)
                    meepo["itemHandle"]=itemHandle
                    meepo["item"]=item
                end
            else
                meepo["itemHandle"]=meepo.AddItemByName(meepo,item)
                meepo["item"]=item
            end
        end
        for j=0,5,1 do
            local itemToCheck = meepo.GetItemInSlot(meepo,j)
            if itemToCheck then
                local name = itemToCheck.GetAbilityName(itemToCheck)
                if name~=item then
                    UTIL_Remove(itemToCheck)
                end
            end
        end
        meepo.SetBaseStrength(meepo,mainMeepo.GetStrength(mainMeepo))
        meepo.SetBaseAgility(meepo,mainMeepo.GetAgility(mainMeepo))
        meepo.SetBaseIntellect(meepo,mainMeepo.GetIntellect(mainMeepo))
        meepo.CalculateStatBonus(meepo)
        while meepo.GetLevel(meepo)<mainMeepo.GetLevel(mainMeepo) do
            meepo.AddExperience(meepo,10,1,false,false)
        end
    else
        LevelAbilitiesForAllMeepos(meepo)
    end
end
function modifier_meepo_divided_we_stand_lua.OnTakeDamage(self,keys)
    local mainMeepo = self.GetCaster(self)
    local parent = self.GetParent(self)
    if keys.unit==parent then
        if parent.GetHealth(parent)<0 then
            if parent~=self.GetCaster(self) then
                local oldLocation = mainMeepo.GetAbsOrigin(mainMeepo)
                mainMeepo.SetAbsOrigin(mainMeepo,parent.GetAbsOrigin(parent))
                mainMeepo.Kill(mainMeepo,keys.inflictor,keys.attacker)
                mainMeepo.SetAbsOrigin(mainMeepo,oldLocation)
                parent.SetHealth(parent,parent.GetMaxHealth(parent))
            end
            for _, meepo in pairs(GetAllMeepos(mainMeepo)) do
                if meepo~=mainMeepo then
                    meepo.AddNewModifier(meepo,mainMeepo,self.GetAbility(self),"modifier_meepo_divided_we_stand_death_lua",{})
                end
            end
        end
    end
end
function modifier_meepo_divided_we_stand_lua.OnRespawn(self,keys)
    local parent = self.GetParent(self)
    local mainMeepo = self.GetCaster(self)
    if (keys.unit==parent) and (parent==PlayerResource.GetSelectedHeroEntity(PlayerResource,self.GetParent(self).GetPlayerOwnerID(self.GetParent(self)))) then
        for _, meepo in pairs(GetAllMeepos(mainMeepo)) do
            if meepo~=mainMeepo then
                meepo.RemoveModifierByName(meepo,"modifier_meepo_divided_we_stand_death_lua")
                meepo.RemoveNoDraw(meepo)
                FindClearSpaceForUnit(meepo,self.GetParent(self).GetAbsOrigin(self.GetParent(self)),true)
                meepo.AddNewModifier(meepo,meepo,self.GetAbility(self),"modifier_phased",{["duration"]=0.1})
            end
        end
    end
end
modifier_meepo_divided_we_stand_death_lua = class({})
function modifier_meepo_divided_we_stand_death_lua.IsPermanent(self)
    return false
end
function modifier_meepo_divided_we_stand_death_lua.IsHidden(self)
    return true
end
function modifier_meepo_divided_we_stand_death_lua.OnCreated(self)
    if IsServer() then
        self.GetParent(self).StartGesture(self.GetParent(self),ACT_DOTA_DIE)
        self.StartIntervalThink(self,1.5)
    end
end
function modifier_meepo_divided_we_stand_death_lua.OnIntervalThink(self)
    local parent = self.GetParent(self)
    parent.RemoveGesture(parent,ACT_DOTA_DIE)
    parent.AddNoDraw(parent)
    if parent.GetTeamNumber(parent)==DOTA_TEAM_GOODGUYS then
        parent.SetAbsOrigin(parent,Vector(-10000,-10000,0))
    else
        parent.SetAbsOrigin(parent,Vector(10000,10000,0))
    end
    self.StartIntervalThink(self,-1)
end
function modifier_meepo_divided_we_stand_death_lua.CheckState(self)
    return {[MODIFIER_STATE_STUNNED]=true,[MODIFIER_STATE_UNSELECTABLE]=true,[MODIFIER_STATE_INVULNERABLE]=true,[MODIFIER_STATE_UNTARGETABLE]=true,[MODIFIER_STATE_OUT_OF_GAME]=true,[MODIFIER_STATE_NO_HEALTH_BAR]=true,[MODIFIER_STATE_NOT_ON_MINIMAP]=true}
end
function LevelAbilitiesForAllMeepos(caster)
    local PID = caster.GetPlayerOwnerID(caster)
    local mainMeepo = PlayerResource.GetSelectedHeroEntity(PlayerResource,PID)
    if caster==mainMeepo then
        for a=0,caster.GetAbilityCount(caster)-1,1 do
            local ability = caster.GetAbilityByIndex(caster,a)
            if ability then
                for _, meepo in pairs(GetAllMeepos(mainMeepo)) do
                    local cloneAbility = meepo.FindAbilityByName(meepo,ability.GetAbilityName(ability))
                    if ability.GetLevel(ability)>cloneAbility.GetLevel(cloneAbility) then
                        cloneAbility.SetLevel(cloneAbility,ability.GetLevel(ability))
                        meepo.SetAbilityPoints(meepo,meepo.GetAbilityPoints(meepo)-1)
                    end
                    if ability.GetLevel(ability)<cloneAbility.GetLevel(cloneAbility) then
                        ability.SetLevel(ability,cloneAbility.GetLevel(cloneAbility))
                        mainMeepo.SetAbilityPoints(mainMeepo,mainMeepo.GetAbilityPoints(mainMeepo)-1)
                    end
                end
            end
        end
    end
end
function GetAllMeepos(caster)
    if caster.meepoList then
        return caster.meepoList
    else
        return {caster}
    end
end
function MeepoExperience(filterTable)
    local PID = filterTable.player_id_const
    local reason = filterTable.reason_const
    local experience = filterTable.experience
    local hero = PlayerResource.GetSelectedHeroEntity(PlayerResource,PID)
    if (hero and hero.HasAbility(hero,"meepo_divided_we_stand_lua")) and (reason~=DOTA_ModifyXP_Unspecified) then
        filterTable.experience=0
        hero.AddExperience(hero,experience,DOTA_ModifyXP_Unspecified,true,false)
    end
    return filterTable
end
function MeepoOrderFilter(filterTable)
    local entindex_ability = filterTable.entindex_ability
    local sequence_number_const = filterTable.sequence_number_const
    local queue = filterTable.sequence_number_const
    local units = filterTable.units
    local entindex_target = filterTable.entindex_target
    local position = Vector(filterTable.position_y,filterTable.position_y,filterTable.position_z)
    local order_type = filterTable.order_type
    local issuer_player_id_const = filterTable.issuer_player_id_const
    local ability = EntIndexToHScript(entindex_ability)
    local target = EntIndexToHScript(entindex_target)
    for _, entindex_unit in pairs(units) do
        local unit = EntIndexToHScript(entindex_unit)
        if unit.HasModifier(unit,"modifier_meepo_divided_we_stand_lua") and (unit~=PlayerResource.GetSelectedHeroEntity(PlayerResource,unit.GetPlayerOwnerID(unit))) then
            if order_type==DOTA_UNIT_ORDER_PICKUP_ITEM then
                return false
            end
        end
        if ((target and target.HasModifier) and target.HasModifier(target,"modifier_meepo_divided_we_stand_lua")) and (target~=PlayerResource.GetSelectedHeroEntity(PlayerResource,target.GetPlayerOwnerID(target))) then
            if order_type==DOTA_UNIT_ORDER_GIVE_ITEM then
                return false
            end
        end
    end
    return true
end
if IsServer() then
    GameRules.GetGameModeEntity(GameRules).SetExecuteOrderFilter(GameRules.GetGameModeEntity(GameRules),function(f,filterTable)
        if MeepoOrderFilter then
            if not MeepoOrderFilter(filterTable) then
                return false
            end
        end
        return true
    end ,GameRules.GetGameModeEntity(GameRules))
    GameRules.GetGameModeEntity(GameRules).SetModifyExperienceFilter(GameRules.GetGameModeEntity(GameRules),function(f,filterTable)
        if MeepoExperience then
            filterTable=MeepoExperience(filterTable)
        end
        return true
    end ,GameRules.GetGameModeEntity(GameRules))
end
